nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/src/main/java/org/apache/nifi/processors/hive/PutHive3Streaming.java

```diff
@@ -0,0 +1,548 @@
+
+        final ComponentLog log = getLogger();
+        final String metastoreUri = context.getProperty(METASTORE_URI).evaluateAttributeExpressions(flowFile).getValue();
+
+        final String partitionValuesString = context.getProperty(PARTITION_VALUES).evaluateAttributeExpressions(flowFile).getValue();
+        final boolean autoCreatePartitions = context.getProperty(AUTOCREATE_PARTITIONS).asBoolean();
+        final boolean disableStreamingOptimizations = context.getProperty(DISABLE_STREAMING_OPTIMIZATIONS).asBoolean();
+
+        HiveOptions o = new HiveOptions(metastoreUri, dbName, tableName)
+                .withHiveConf(hiveConfig)
+                .withAutoCreatePartitions(autoCreatePartitions)
+                .withCallTimeout(callTimeout)
+                .withStreamingOptimizations(!disableStreamingOptimizations);
+
+        if (!StringUtils.isEmpty(partitionValuesString)) {
+            List<String> staticPartitionValues = Arrays.stream(partitionValuesString.split(",")).filter(Objects::nonNull).map(String::trim).collect(Collectors.toList());
+            o = o.withStaticPartitionValues(staticPartitionValues);
+        }
+
+        if (SecurityUtil.isSecurityEnabled(hiveConfig)) {
+            final KerberosCredentialsService credentialsService = context.getProperty(KERBEROS_CREDENTIALS_SERVICE).asControllerService(KerberosCredentialsService.class);
+            o = o.withKerberosPrincipal(credentialsService.getPrincipal()).withKerberosKeytab(credentialsService.getKeytab());
+        }
+
+        final HiveOptions options = o;
+
+        // Store the original class loader, then explicitly set it to this class's classloader (for use by the Hive Metastore)
+        ClassLoader originalClassloader = Thread.currentThread().getContextClassLoader();
+        Thread.currentThread().setContextClassLoader(this.getClass().getClassLoader());
+
+        StreamingConnection hiveStreamingConnection = null;
+
+        try (final InputStream rawIn = session.read(flowFile)) {
+            long processedRecords = 0L;
+            final RecordReader reader;
+
+            try (final BufferedInputStream in = new BufferedInputStream(rawIn)) {
+
+                // if we fail to create the RecordReader then we want to route to failure, so we need to
+                // handle this separately from the other IOExceptions which normally route to retry
+                try {
+                    reader = recordReaderFactory.createRecordReader(flowFile, in, getLogger());
+                } catch (Exception e) {
+                    throw new RecordReaderFactoryException("Unable to create RecordReader", e);
+                }
+
+                hiveStreamingConnection = makeStreamingConnection(options, reader);
+
+                // Write records to Hive streaming, then commit and close
+                hiveStreamingConnection.beginTransaction();
+                hiveStreamingConnection.write(in);
```

Comment:
Looks like we are creating one connection per flow file. I don't know anything about flow file. How many rows does flow file typically have? Recommendation is to avoid making too frequent commits. Commit after 100000 to 1M rows will get best performance.

Reply:
We'll encourage users of this processor to have flow files with a significant number of records, although I don't think the average use case will have that many rows. Certainly we do not want to commit a single row at a time, but we'd need special logic to keep flow files (and the connection) around until 1M rows is achieved, because we want to wait until the rows are committed before we send the flow files downstream. This is how MergeContent and MergeRecord work, we are suggesting that users combine smaller flow files into larger ones before sending to PutHive3Streaming

Reply:
Sounds good.